File: server/Services/RecipeService.cs
namespace wk10.Services;

public class RecipeService(RecipeRepo recipeRepo)
{
  internal List<Recipe> GetRecipes()
  { return recipeRepo.GetRecipes(); }

  internal Recipe GetRecipeById(int recipeId)
  { return recipeRepo.GetRecipeById(recipeId) ?? throw new Exception($"Recipe with ID {recipeId} does not exist"); }

  internal Recipe CreateRecipe(Recipe recipeData)
  { return recipeRepo.CreateRecipe(recipeData); }

  internal string DeleteRecipe(string creatorId, int recipeId)
  {
    Recipe recipe = GetRecipeById(recipeId);
    if (recipe.CreatorId != creatorId) { throw new Exception("Forbidden action - Not your data"); }
    recipeRepo.DeleteRecipe(recipeId);
    return "Recipe deleted";
  }

  internal Recipe UpdateRecipe(int recipeId, Recipe recipeData)
  {
    Recipe recipe = GetRecipeById(recipeId);

    recipe.Title = recipeData.Title ?? recipe.Title;
    recipe.Subtitle = recipeData.Subtitle ?? recipe.Subtitle;
    recipe.Instructions = recipeData.Instructions ?? recipe.Instructions;
    recipe.Category = recipeData.Category ?? recipe.Category;
    recipe.Img = recipeData.Img ?? recipe.Img;

    recipeRepo.UpdateRecipe(recipe);

    return recipe;
  }


}

File: client/src/services/RecipeService.js
import { AppState } from "../AppState";
import { Ingredient } from "../models/Ingredient";
import { Recipe } from "../models/Recipe";
import { logger } from "../utils/Logger";
import { api } from "./AxiosService";

function recipeIndex(recipeId) {
  const index = AppState.recipes.findIndex(r => r.id == recipeId);
  if (index == -1) { throw new Error('Unable to locate recipe') }
  return index
}
class RecipeService {

  async getRecipes() {
    const res = await api.get('api/recipes');
    AppState.recipes = res.data.map(r => new Recipe(r));
  }

  async getRecipeById(recipeId) {
    const res = await api.get('api/recipes/' + recipeId);
    const recipe = new Recipe(res.data);
    AppState.activeRecipe = recipe; // overwrite in new copy in case of changes since page load
    const index = recipeIndex(recipeId)
    AppState.recipes.splice(index, 1, recipe) // splice in new copy in case of changes since page load
  }
  async getIngredientsByRecipeId(recipeId) {
    const res = await api.get('api/recipes/' + recipeId + '/ingredients');
    AppState.ingredients = res.data.map(i => new Ingredient(i));
  }

  async createRecipe(recipeData) {
    const res = await api.post('api/recipes', recipeData);
    AppState.recipes.push(new Recipe(res.data));
  }

  async deleteRecipe(recipeId) {
    const res = await api.delete('api/recipes/' + recipeId);
    AppState.recipes = AppState.recipes.filter(r => r.id != recipeId);
    logger.log(`Recipe with ID ${recipeId} has been deleted`)
    AppState.activeRecipe = {};
  }

  async updateRecipe(recipeData) {
    const res = await api.put('api/recipes/' + recipeData.id, recipeData);
    const index = recipeIndex(recipeData.id)
    AppState.recipes.splice(index, 1, recipeData);
  }
}

export const recipeService = new RecipeService();